File: src/main/scala/booklet/impure/router/BookFinderRouter.scala
package booklet.impure.router

import booklet.impure.service.bookfinder.GoogleBookFinder
import booklet.pure.http.CustomResponse.{badRequest, serverFailure}
import booklet.pure.http.Query
import zhttp.http.Method.GET
import zhttp.http._
import zio.ZIO

object BookFinderRouter {
  val app: Http[GoogleBookFinder, Nothing, Request, Response] =
    Http.collectZIO[Request] { case req @ GET -> !! / "find" =>
      Query.param(req)(name = "isbn") match {
        case None => ZIO.succeed(badRequest("Missing ISBN"))
        case Some(isbn) =>
          GoogleBookFinder
            .findByIsbn(isbn)
            .fold(
              serverFailure,
              {
                case None       => badRequest(s"No book found for ISBN $isbn")
                case Some(book) => Response.text(book.toString)
              }
            )
      }
    }
}

File: src/main/ts/booklet/impure/router/BookFinderRouter.ts
import { Http, Request, Response } from 'zhttp/http';
import { GoogleBookFinder } from '../service/GoogleBookFinder';
import { Failure } from '../../pure/Failure';
import { CustomResponse } from '../../pure/http/CustomResponse';

export const BookFinderRouter = {
  app: Http.collectZIO<Request, Response>({
    'GET /books/find': async (req: Request) => {
      const isbn = req.query.isbn;
      const idToken = req.headers['Authorization'];
      if (!isbn || !idToken) {
        return CustomResponse.badRequest('Missing ISBN or ID token');
      }
      try {
        const book = await GoogleBookFinder.findByIsbn(isbn, idToken);
        if (book) {
          return CustomResponse.ok(book);
        } else {
          return CustomResponse.notFound();
        }
      } catch (error) {
        return CustomResponse.serverFailure(Failure.fromThrowable(error));
      }
    }
  })
};
